consensus/misc/forks.go

```diff
@@ -41,3 +52,72 @@ func VerifyForkHashes(config *params.ChainConfig, header *types.Header, uncle bo
 	// All ok, return
 	return nil
 }
+
+// VerifyEIP1559BaseFee verifies that the EIP1559 BaseFee field is valid for the current block height
+func VerifyEIP1559BaseFee(config *params.ChainConfig, header, parent *types.Header) error {
+	// If we are at the EIP1559 fork block the BaseFee needs to be equal to params.EIP1559InitialBaseFee
+	if config.EIP1559Block != nil && config.EIP1559Block.Cmp(header.Number) == 0 {
+		if header.BaseFee == nil || header.BaseFee.Cmp(new(big.Int).SetUint64(params.EIP1559InitialBaseFee)) != 0 {
+			return errInvalidInitialBaseFee
+		}
+		return nil
+	}
+	// Verify the BaseFee is valid if we are past the EIP1559 activation block
+	if config.IsEIP1559(header.Number) {
+		if parent.BaseFee == nil {
+			return errMissingParentBaseFee
+		}
+		if header.BaseFee == nil {
+			return errMissingBaseFee
+		}
+		delta := new(big.Int).Sub(new(big.Int).SetUint64(parent.GasUsed), new(big.Int).SetUint64(params.TargetGasUsed))
```

Comment:
Hello,
I am wondering there is an issue on the delta computation. The EIP says `Let delta = block.gas_used - TARGET_GASUSED`.
I would expect to have `header.GasUsed` instead of `parent.GasUsed` . Am i missing something ?

Reply:
Thanks for bringing this up, this highlights an issue in the spec. It should clarify that we are using the `delta` of the current header to calculate the `baseFee` of the next header; the parent to calculate the current. So in this function we are confirming that the `baseFee` provided in a header is valid by deriving it from its parent using the consensus rules and checking that they are equal.

Miners use the `baseFee` of the header for the block they are working on to calculate the `gasPrice` of the EIP1559 txs it includes, so a header needs its `baseFee` value known before its `gasUsed` value is resolved.

Reply:
> Thanks for bringing this up, this highlights an issue in the spec. It should clarify that we are using the `delta`/`gasUsed` of the current header to calculate the `baseFee` of the next header; the parent to calculate the current. So in this function we are confirming that the `baseFee` provided in a header is valid by deriving it from its parent using the consensus rules and checking that they are equal.
> 
> Currently, miners use the `baseFee` of the header for the block they are working on to calculate the `gasPrice` of the EIP1559 txs they include, so a header needs its `baseFee` value known before its `gasUsed` value is resolved.

Thanks for the clarification 👍